text/0070-pd-record-historical-hot-region.md

```diff
@@ -0,0 +1,202 @@
+  | 328              | 10                   | 45.04394531              | 1351.318359                |
+  | 328              | 15                   | 30.02929688              | 900.8789063                |
+
+### Design in PD
+1. Timing write：
+
+     The leader of PD will periodically encrypt  `START_KEY` and `END_KEY`  in data,and write hotspot region data into `LevelDB`.The write interval can be configured. The write fieldes are: `REGION_ID`, `TYPE`,  `HOT_DEGREE`, `FLOW_BYTES`, `KEY_RATE`, `QUERY_RATE`, `STORE_ID`, `PEER_ID`, `UPDATE_TIME`, `START_KEY`,  `END_KEY`.
+
+2. Timing delete
+
+   PD runs the delete check task periodically, and deletes the hotspot region data that exceeds the configured TTL time.
+
+3. Pull interface
+
+   PD query the data in `LevleDB` with filters pushed down, decodes the data and returns to TiDB.
+
+4. New options
+
+   There are two config options need to be add in PD’s `config.go`:
+
+   * `HisHotRegionSaveInterval`:  time interval for pd to record hotspot region information, default: 15 minutes.
+   * `HisHotRegionTTL`: maximum hold day for his hot region, default: 30 days.
+
+### Design in TiDB
+
+1. Add memory table `TIDB_HOT_REGIONS_HISTORY`：
+
+   Create a new memory table `TIDB_HOT_REGIONS_HISTORY` with fileds discussed above in `INFORMATION_SCHEMA` schema.
+
+1. Add `HotRegionsHistoryTableExtractor` to push down some predicates to PD in order to  reduce network IO.
+
+   ```go
+   type HistoryHotRegionsRequest struct {
+   	StartTime      int64    `json:"start_time,omitempty"`
+   	EndTime        int64    `json:"end_time,omitempty"`
+   	RegionIDs      []uint64 `json:"region_ids,omitempty"`
+   	StoreIDs       []uint64 `json:"store_ids,omitempty"`
+   	PeerIDs        []uint64 `json:"peer_ids,omitempty"`
+   	HotRegionTypes []string `json:"hot_region_types,omitempty"`
+   	LowHotDegree   int64    `json:"low_hot_degree,omitempty"`
+   	HighHotDegree  int64    `json:"high_hot_degree,omitempty"`
+   	LowFlowBytes   float64  `json:"low_flow_bytes,omitempty"`
+   	HighFlowBytes  float64  `json:"high_flow_bytes,omitempty"`
+   	LowKeyRate     float64  `json:"low_key_rate,omitempty"`
+   	HighKeyRate    float64  `json:"high_key_rate,omitempty"`
+   	LowQueryRate   float64  `json:"low_query_rate,omitempty"`
+   	HighQueryRate  float64  `json:"high_query_rate,omitempty"`
+   }
+   ```
+   
+1. Add `hotRegionsHistoryRetriver` to fetch hotspot regions from all  PD servers by HTTP request,  then supplement fields like `DB_NAME`, `TABLE_NAME`,  `TABLE_ID`,  `INDEX_NAME`, `INDEX_ID` according to the `START_KEY` and `END_KEY`of the hotspot region, and merge the results.
+
+## DrawBack
+
+* TiDB can not add fields for deleted Table or Schema.
+
+## Alternatives
+
+Write to TiDB like a normal table.
+* Reuse complete push-down function.
+* Data is written into TiKV to provide disaster tolerance.
+* The `Owner` election solves the problem of who pulls data from PD.
+* Can not support scenarios using TiKV independently.
+
```

Comment:
What about creating a new tool to periodically pull data from pd and insert them into TiDB or other log services, instead of making this into pd?

Reply:
If the tool is in TiDB, it is similay with alternative we mention in `Alternatives`, one disadvantage is that it can not support pure TiKV scenario.  As for independent tool like [ecosystem-tool](https://docs.pingcap.com/zh/tidb/stable/ecosystem-tool-user-guide),  we'll talk more about that.